[Concurrency] Allow getting unsafe task from Job and name as well

ktoso opened the pull request:
**Explanation**: 
This was discussed in https://github.com/swiftlang/swift-evolution/blob/main/proposals/0469-task-names.md#unsafecurrenttask-access-from-unownedjob which was accepted however this part wasn't actually implemented it seems.


This change required some changes in how we handle records, since the
name must be FIRST on the stack right now, since it can outlive the task
executor preference. Added some tests

**Scope**: Narrow, adds member `task.name` and `unsafeCurrentTask` to `ExecutorJob` which were previously approved in SE, but missing from implementation.

**Risk**: Low, adds new APIs, all 6.4 gated, without backdeployment as we cannot backdeploy the runtime function.

**Testing**: 
Added CI tests reading the names using the new APIs

**Issues**: 
Resolves rdar://157558738

ktoso commented:
@swift-ci please smoke test

FranzBusch commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -611,6 +611,11 @@ extension Task where Success == Never, Failure == Never {
+
+  @available(SwiftStdlib 6.2, *)
+  public var name: String? {
+    return _getCurrentTaskNameString()
```
This shouldn't call `getCurrentTaskName` but `_getTaskName` right?

ktoso replied:
thanks yep :) also missed to push tests, added now 😅

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test Linux

ktoso commented:
@swift-ci please smoke test Windows

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test

ktoso committed:
[Concurrency] Allow getting unsafe task from Job and name as well

Don't add unsafeCurrentTask for the time being

adjust comments and make @inlinable the property names

ktoso commented:
@swift-ci please smoke test

ktoso commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -630,6 +630,29 @@ extension Task where Success == Never, Failure == Never {
+  /// however at runtime, this function is only able to return actual value Swift runtime versions `6.4+`.
+  /// On older runtimes, the task name will return `nil`.
+  @available(SwiftStdlib 6.2, *)
+  public var name: String? {
```
We can make this backdeployed but only "work" on 6.4, this is similar to what we did with.

This matches the availability of the static `Task.name`

ktoso commented:
@swift-ci please smoke test

ktoso commented:
The "unrelated" changes are https://github.com/swiftlang/swift/pull/88375 which should be merged by merge wrangler tonight.

ktoso commented:
@swift-ci please smoke test

rjmccall commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -876,12 +890,54 @@ public struct UnsafeCurrentTask {
+    let task = unsafe unsafeBitCast(self.context, to: Builtin.NativeObject.self)
+    return unsafe UnsafeCurrentTask(task)
+  }
+}
```
What is this doing in this patch? This wasn't an approved change, and it's specifically quite weird (and dangerous?) — `UnsafeCurrentTask` is an API surface for querying the *current* task, not an unsafe API for accessing arbitrary tasks.

ktoso replied:
As the opening comment says: https://github.com/swiftlang/swift-evolution/blob/main/proposals/0469-task-names.md#unsafecurrenttask-access-from-unownedjob 

this *was* actually in the original proposal that was accepted.

do we want to not have this even though it went through SE? It was meant to access the names of jobs.

ktoso replied:
Discussed and removing it for now from this PR, we'll revisit the accessor

rjmccall replied:
Thanks, looks good.

ktoso commented:
Heh, noticing all the comments I needed to adjust took me a few rounds. Good now, please review

ktoso commented:
@swift-ci please smoke test

ktoso commented:
Unrelated swift-testing failure

```
  TestContentGeneration.swift:21:72: error: cannot convert value of type 'String'
    to expected argument type 'String.Element' (aka 'Character')
   19 |   /// not.
   20 |   fileprivate var commentRepresentation: Trivia {
   21 |     guard let fourCharacterCodeValue, !fourCharacterCodeValue.contains("*/") else {
      |                                                                        ^
      |                        error: cannot convert value of type 'String' to
      |                        expected argument type 'String.Element' (aka 'Character')
   22 |       return []
```

ktoso commented:
@swift-ci please smoke test Linux platform

ktoso commented:
@swift-ci please test Windows platform

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test Linux platform

ktoso committed:
The name must also be before any copied-from-parent records

ktoso commented:
Actually still had a bug here if we were copying values from parent task the name would not be first. Fixed now

ktoso commented:
@swift-ci please smoke test

ktoso commented:
Found another issue, we must have the name be before the parent link on the stack, otherwise we can run into issues as we destroy "all but one" but the "but one" is the parent link in task locals, and not the record.

ktoso committed:
Move task record EVEN EARLIER before the parent link of task local

ktoso commented:
@swift-ci please smoke test
